### ISSUE TITLE
[Bug]: Sharing a directory received via federation share doesn't work

### ISSUE THREAD

**Issue by Bibo-Joshi** [nextcloud/server#36320]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

When receiving a directory share via federiation share with full permissions (especially "allow to re-share"), creating share-links for the doesn't work. Instead, an error message shows in the browser reading

> Error while creating the share

In the logs, I find the following traceback:

<details>

```json

{
    "reqId": "9OLlQOLINeXsTD5L6wP6",
    "level": 3,
    "time": "2023-01-23T20:55:17+00:00",
    "remoteAddr": "123",
    "user": "xxx",
    "app": "no app in context",
    "method": "POST",
    "url": "/ocs/v2.php/apps/files_sharing/api/v1/shares",
    "message": "call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method \"getShareId\" in file '/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php' line 527",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 OPR/94.0.0.0",
    "version": "25.0.3.2",
    "exception": {
        "Exception": "Exception",
        "Message": "call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method \"getShareId\" in file '/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php' line 527",
        "Code": 0,
        "Trace": [
            {
                "file": "/media/webspace/nextcloud/lib/private/AppFramework/App.php",
                "line": 172,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
            },
            {
                "file": "/media/webspace/nextcloud/lib/private/Route/Router.php",
                "line": 298,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
            },
            {
                "file": "/media/webspace/nextcloud/ocs/v1.php",
                "line": 63,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
            },
            {
                "file": "/media/webspace/nextcloud/ocs/v2.php",
                "line": 23,
                "args": ["/media/webspace/nextcloud/ocs/v1.php"],
                "function": "require_once",
            },
        ],
        "File": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "Line": 165,
        "Previous": {
            "Exception": "TypeError",
            "Message": 'call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method "getShareId"',
            "Code": 0,
            "Trace": [
                {
                    "file": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
                    "line": 527,
                    "function": "call_user_func_array",
                },
                {
                    "function": "__call",
                    "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
                    "line": 527,
                    "function": "call_user_func_array",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Share20/Manager.php",
                    "line": 677,
                    "function": "__call",
                    "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Share20/Manager.php",
                    "line": 767,
                    "function": "setLinkParent",
                    "class": "OC\\Share20\\Manager",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/apps/files_sharing/lib/Controller/ShareAPIController.php",
                    "line": 701,
                    "function": "createShare",
                    "class": "OC\\Share20\\Manager",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 225,
                    "function": "createShare",
                    "class": "OCA\\Files_Sharing\\Controller\\ShareAPIController",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 133,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/App.php",
                    "line": 172,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Route/Router.php",
                    "line": 298,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::",
                },
                {
                    "file": "/media/webspace/nextcloud/ocs/v1.php",
                    "line": 63,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/ocs/v2.php",
                    "line": 23,
                    "args": ["/media/webspace/nextcloud/ocs/v1.php"],
                    "function": "require_once",
                },
            ],
            "File": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
            "Line": 527,
        },
        "CustomMessage": "--",
    },
    "id": "63cef45e0ccd7",
}
```

</details>

### Steps to reproduce

1. Share a directory from NC A to user John on NC B, set all permissions for the share
2. as user John on NC B, try to create a share link for the received directory
3. See error message in the browser & in the logs

### Expected behavior

With the "allow to re-share" permission, I can share the directory just like I can do with received non-federation shares or my own directories.

### Installation method

Community Manual installation with Archive

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "mahlerhome.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "http:\/\/mahlerhome.de",
        "dbtype": "mysql",
        "version": "25.0.3.2",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "maintenance": false,
        "loglevel": 0,
        "theme": "",
        "updater.release.channel": "stable",
        "app_install_overwrite": [
            "mail",
            "bruteforcesettings",
            "files_downloadactivity",
            "files_trackdownloads",
            "apporder",
            "documentserver_community",
            "files_markdown",
            "onlyoffice",
            "polls",
            "twofactor_totp",
            "impersonate"
        ],
        "log_rotate_size": 10485760,
        "default_phone_region": "DE",
        "mail_sendmailmode": "smtp"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - apporder: 0.15.0
  - bruteforcesettings: 2.5.0
  - calendar: 4.2.1
  - cloud_federation_api: 1.8.0
  - contacts: 5.0.2
  - contactsinteraction: 1.6.0
  - dav: 1.24.0
  - documentserver_community: 0.1.13
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_downloadactivity: 1.15.0
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trackdownloads: 1.11.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - forms: 3.0.3
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - mail: 2.2.2
  - music: 1.8.1
  - nextcloud_announcements: 1.14.0
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - onlyoffice: 7.6.8
  - password_policy: 1.15.0
  - photos: 2.0.1
  - polls: 4.0.0
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - related_resources: 1.0.3
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - support: 1.8.0
  - suspicious_login: 4.3.0
  - tasks: 0.14.5
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_backupcodes: 1.14.0
  - twofactor_totp: 7.0.0
  - updatenotification: 1.15.0
  - viewer: 1.9.0
  - workflowengine: 2.7.0
Disabled:
  - circles: 23.0.0
  - comments: 1.13.0
  - dashboard: 7.0.0
  - encryption
  - files_external
  - impersonate: 1.11.0
  - recommendations: 0.4.0
  - survey_client: 1.9.0
  - systemtags: 1.13.0
  - user_ldap
  - user_status: 1.1.1
  - weather_status: 1.3.0
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
{
    "reqId": "9OLlQOLINeXsTD5L6wP6",
    "level": 3,
    "time": "2023-01-23T20:55:17+00:00",
    "remoteAddr": "123",
    "user": "xxx",
    "app": "no app in context",
    "method": "POST",
    "url": "/ocs/v2.php/apps/files_sharing/api/v1/shares",
    "message": "call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method \"getShareId\" in file '/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php' line 527",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36 OPR/94.0.0.0",
    "version": "25.0.3.2",
    "exception": {
        "Exception": "Exception",
        "Message": "call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method \"getShareId\" in file '/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php' line 527",
        "Code": 0,
        "Trace": [
            {
                "file": "/media/webspace/nextcloud/lib/private/AppFramework/App.php",
                "line": 172,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
            },
            {
                "file": "/media/webspace/nextcloud/lib/private/Route/Router.php",
                "line": 298,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
            },
            {
                "file": "/media/webspace/nextcloud/ocs/v1.php",
                "line": 63,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
            },
            {
                "file": "/media/webspace/nextcloud/ocs/v2.php",
                "line": 23,
                "args": ["/media/webspace/nextcloud/ocs/v1.php"],
                "function": "require_once",
            },
        ],
        "File": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "Line": 165,
        "Previous": {
            "Exception": "TypeError",
            "Message": 'call_user_func_array(): Argument #1 ($callback) must be a valid callback, class OCA\\Files_Sharing\\External\\Storage does not have a method "getShareId"',
            "Code": 0,
            "Trace": [
                {
                    "file": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
                    "line": 527,
                    "function": "call_user_func_array",
                },
                {
                    "function": "__call",
                    "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
                    "line": 527,
                    "function": "call_user_func_array",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Share20/Manager.php",
                    "line": 677,
                    "function": "__call",
                    "class": "OC\\Files\\Storage\\Wrapper\\Wrapper",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Share20/Manager.php",
                    "line": 767,
                    "function": "setLinkParent",
                    "class": "OC\\Share20\\Manager",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/apps/files_sharing/lib/Controller/ShareAPIController.php",
                    "line": 701,
                    "function": "createShare",
                    "class": "OC\\Share20\\Manager",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 225,
                    "function": "createShare",
                    "class": "OCA\\Files_Sharing\\Controller\\ShareAPIController",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 133,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/AppFramework/App.php",
                    "line": 172,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/lib/private/Route/Router.php",
                    "line": 298,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::",
                },
                {
                    "file": "/media/webspace/nextcloud/ocs/v1.php",
                    "line": 63,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->",
                },
                {
                    "file": "/media/webspace/nextcloud/ocs/v2.php",
                    "line": 23,
                    "args": ["/media/webspace/nextcloud/ocs/v1.php"],
                    "function": "require_once",
                },
            ],
            "File": "/media/webspace/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php",
            "Line": 527,
        },
        "CustomMessage": "--",
    },
    "id": "63cef45e0ccd7",
}
```

### Additional info

Follow up of #33190

**Comment by Hurri08** [comment]
Still having the same Problem with both instances on Nextcloud 26.01 and PHP-8.1. Is there any progress or an eta for a possible fix?

**Comment by andreas-bulling** [comment]
Same problem here - why is nobody at Nextcloud fixing this? This seems to affect quite a number of people and it breaks a core functionality.

**Comment by joshtrichards** [comment]
Fixed in #47038
